### PR TITLE
Friends App

### PR DESCRIPTION
**Pull request by rrrds**
### Friends App

[Demo](https://rrrds.github.io/friends-app/index.html) |
[Code base](https://github.com/rrrds/friends-app)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/rrrds/friends-app/js/app.js [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    this.setState(this.config.initialState || {});
+
+    this.components.forEach(c => {
+      c.setValue(this.state);
+    });
+  }
+
+  fetchData(apiUrl) {
+    return fetch(apiUrl)
+      .then(response => response.json())
```
Comment (zonzujiro) [inline submissions/rrrds/friends-app/js/app.js]: https://www.tjvantoll.com/2015/09/13/fetch-and-errors/

File: submissions/rrrds/friends-app/js/app.js [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    return data;
+  }
+
+  renderUsers(data) {
+    this.container.innerHTML = '';
+    const userTemplates = [];
+
+    data.forEach(element => {
```
Comment (zonzujiro) [inline submissions/rrrds/friends-app/js/app.js]: `Array#map` maybe?

File: submissions/rrrds/friends-app/js/app.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+    this.container = document.querySelector('.users-container__body');
+    this.totalCounter = document.querySelector('.users-container__total');
+    this.usersData = [];
+
+    this.config.apiUrl = api;
+  }
+
+  filterByName(data, filter) {
+    return data.filter(item => item.name.first.startsWith(filter));
```
Comment (zonzujiro) [inline submissions/rrrds/friends-app/js/app.js]: Why only first name and why only `startsWith`?

File: submissions/rrrds/friends-app/js/app.js [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+  }
+
+  bindEvent(key, listener) {
+    this.element.addEventListener(key, listener.bind(this));
+  }
+
+  setValue(state = {}) {
+    if (!this.props.valueName) return;
+
+    if (!Object.prototype.hasOwnProperty.call(state, this.props.valueName)) return;
```
[the remark this answers is not in the document]
Comment (rrrds) [reply submissions/rrrds/friends-app/js/app.js]: I pass full state object as parameter and component updates own value only if state has components property (valueName)

File: submissions/rrrds/friends-app/js/app.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+class NumberControl extends Component {
+  constructor(selector, props) {
+    super(selector, props);
+
+    this.eventKey = 'input';
+    this.bindEvent(this.eventKey, this.eventHandler);
+  }
+
+  eventHandler(event) {
+    const value = event.target.value || event.data || null;
```
[the remark this answers is not in the document]
Comment (rrrds) [reply submissions/rrrds/friends-app/js/app.js]: While playing around with number input I decide to use _event.data_ as fallback.
But now I see that its irrelevant in this case

File: submissions/rrrds/friends-app/js/app.js [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            element.location.city
+          },<br><span class="card__state">${element.location.state}</span></p>
+          <p class="card__phone"><i class="card__phone-icon fas fa-phone"></i> ${element.phone}</p>
+        </div>
+      </div>`
+      );
+    });
+
+    this.container.innerHTML = userTemplates.join('');
+    this.totalCounter.innerHTML = `Total: ${data.length || 0}`;
```
[the remark this answers is not in the document]
Comment (rrrds) [reply submissions/rrrds/friends-app/js/app.js]: Yep, my bad

**Pull request merged**